EIPS/eip-5988.md

```diff
@@ -0,0 +1,163 @@
+
+The `POSEIDON` precompile MUST be available at address `0xPOSEIDON_PRECOMPILE_ADDRESS`.
+
+The precompile MUST be activated at `FORK_BLOCK_NUMBER`.
+
+Here are the Poseidon parameters that the precompile will support:
+
+- `p`: the prime field modulus.
+- `security_level`: the security level measured in bits. Denoted `M` in the Poseidon paper.
+- `alpha`: the power of S-box.
+- `input_rate`: the size of input.
+- `t`: the size of the state.
+- `full_round`: the number of full rounds. Denoted as `R_F` in the Poseidon paper.
+- `partial_round`: the number of partial rounds. Denoted as `R_P` in the Poseidon paper.
+
+The input to the precompile MUST be encoded as follows:
+
+- `p`: 32 bytes
+- `security_level`: 2 bytes
+- `alpha`: 1 byte
+- `input_rate`: 2 bytes
+- `t`: 1 byte
+- `full_round`: 1 byte
+- `partial_round`: 1 byte
+- `input`: `input_rate` \* 32 bytes
+
+```
+[32 bytes for p][2 bytes for security_level][1 byte for alpha][2 bytes for input_rate][1 byte for t][1 byte for full_round][1 byte for partial_round][input_rate * 32 bytes for input]
+```
+
+The precompile should compute the hash function as [specified in the Poseidon paper](https://eprint.iacr.org/2019/458.pdf) and return hash output.
+
+### Example Usage in Solidity
+
+The precompile can be wrapped easily in Solidity to provide a more development-friendly interface to `poseidon_hash` function.
+
+```solidity
+// TODO: Add solidity example
+```
+
+### Gas Costs
+
+```
+TODO: Fill gas costs section and add benchmarks
+```
+
+## Rationale
+
+TODO: Add rationale
+
+## Backwards Compatibility
+
+There is very little risk of breaking backwards-compatibility with this EIP, the sole issue being if someone were to build a contract relying on the address at `0xPOSEIDON_PRECOMPILE_ADDRESS` being empty. The likelihood of this is low, and should specific instances arise, the address could be chosen to be any arbitrary value with negligible risk of collision.
+
+## Test Cases
+
+The Poseidon reference implementation contains test vectors that can be used to test the precompile.
+Those tests are available [here](https://extgit.iaik.tugraz.at/krypto/hadeshash/-/blob/master/code/test_vectors.txt).
+
+## Reference Implementation
+
+The reference implementation for various versions of Starkad and Poseidon can be found [here](https://extgit.iaik.tugraz.at/krypto/hadeshash). The repository also includes [test vectors](https://extgit.iaik.tugraz.at/krypto/hadeshash/-/blob/master/code/test_vectors.txt).
+
+Those test vectors can be used to test the implementation of the precompile in the different client implementations.
+
+TODO: Add initial Geth implementation link
+
+## Security Considerations
+
+Quoting Vitalik Buterin from [Arithmetic hash based alternatives to KZG for proto-danksharding](https://ethresear.ch/t/arithmetic-hash-based-alternatives-to-kzg-for-proto-danksharding-eip-4844/13863) thread:
```

Comment:
Linking to `ethresear.ch` is not likely to be accepted. Please remove this link.

Reply:
sure